src/shared/sleep-config.c

```diff
@@ -209,65 +322,131 @@ static bool battery_discharge_rate_is_valid(int battery_discharge_rate) {
 
-        p = hash_id_discharge_rate;
-        r = extract_many_words(&p, " ", 0, &stored_hash_id, &stored_discharge_rate, NULL);
-        if (r < 0)
-               return log_debug_errno(r, "Failed to parse hash_id and discharge_rate read from %s location: %m", DISCHARGE_RATE_FILEPATH);
-        if (r != 2)
-               return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid number of items fetched from %s", DISCHARGE_RATE_FILEPATH);
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
 
-        r = safe_atou64(stored_hash_id, &hash_id);
-        if (r < 0)
-               return log_debug_errno(r, "Failed to parse discharge rate read from %s location: %m", DISCHARGE_RATE_FILEPATH);
+                r = read_line(f, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read \"%s\": %m", DISCHARGE_RATE_FILEPATH);
+                if (r == 0)
+                        break;
 
-        r = get_system_battery_identifier_hash(&current_hash_id);
-        if (r < 0)
-               return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
+                p = line;
+                r = extract_many_words(&p, " ", 0, &stored_hash_id, &stored_discharge_rate, NULL);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to parse hash_id and discharge_rate read from %s location: %m", DISCHARGE_RATE_FILEPATH);
+                if (r != 2)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid number of items fetched from %s", DISCHARGE_RATE_FILEPATH);
+
+                r = safe_atou64(stored_hash_id, &hash_id);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to parse discharge rate read from %s location: %m", DISCHARGE_RATE_FILEPATH);
+                r = battery_enumerator_new(&e);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to initialize battery enumerator: %m");
 
-        if(current_hash_id != hash_id)
-               return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Current identifier does not match stored identifier: %m");
+                FOREACH_DEVICE(e, dev) {
+                        r = get_system_battery_identifier_hash(dev, &current_hash_id);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to generate system battery identifier hash: %m");
 
-        r = safe_atoi(stored_discharge_rate, &discharge_rate);
-        if (r < 0)
-               return log_debug_errno(r, "Failed to parse discharge rate read from %s location: %m", DISCHARGE_RATE_FILEPATH);
+                        if(current_hash_id != hash_id)
+                                /* matching device not found, move to next device */
+                                continue;
```

Comment:
hmm, please split out this loop into a new helper call: `find_power_supply_by_hash()` or so, that you pass the hash into and that returns you an sd_device object

Reply:
except this and truncate bool, rest all is fixed. will do it today.

Reply:
this is pending

Reply:
it's already a function for the loop basically, apart from the fopen.